### ISSUE TITLE
Upgrade k8s.io/* to v0.23, sigs.k8s.io/controller-runtime to v0.11

### ISSUE THREAD

**Issue by timebertt** [gardener/gardener#5282]
**How to categorize this issue?**

**What would you like to be added**:

We should upgrade to the latest versions of our go upstream dependencies:
- `sigs.k8s.io/controller-runtime@v0.11.x`: https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.11.0
- `k8s.io/*@v0.23.x`
- `sigs.k8s.io/controller-tools@v0.8.x`: https://github.com/kubernetes-sigs/controller-tools/releases/tag/v0.8.0

## Important changes / Action items

Here is a list of a few upstream changes to look out for when vendoring g/g, k/* and c-r in any of our repos (e.g. extensions). Please consider the release notes of [controller-runtime@v0.11.0](https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.11.0) and upwards as well for a more complete list and more details.

### ⚠️ breaking / most notable changes
- Update k8s.io/* dependencies to v0.23.0
- Major version bump of **logr** (v0.4.0 -> v1.2.0)
- Avoid shallow copies of webhooks and CRDs in testenv (https://github.com/kubernetes-sigs/controller-runtime/pull/1667)
### ℹ️ other changes / good to know / for information only:
- https://github.com/kubernetes/kubernetes/pull/94986 A new field `omitManagedFields` has been added to both `audit.Policy` and `audit.PolicyRule` so cluster operators can opt in to omit managed fields of the request and response bodies from being written to the API audit log.
- https://github.com/kubernetes/kubernetes/pull/104983 kube-apiserver's Prometheus metrics have been extended 
  - consider adding metrics for  shoot kube-apiserver ➡️https://github.com/gardener/gardener/pull/5445
- https://github.com/kubernetes/kubernetes/pull/104693  Introduce OS field in the Pod Spec
  - This is an alpha field and requires the IdentifyPodOS feature
- https://github.com/kubernetes-sigs/controller-runtime/pull/1676
  - [x] replace custom coding in seed-admission-controller `extensionresources` webhook (webhook calling validation code for extension resources) with `builder.WebhookBuilder.WithValidator` ➡️ #5782

- https://github.com/kubernetes/kubernetes/pull/104991 included in 1.23 might solve https://github.com/gardener/gardener/issues/4609

**Comment by timebertt** [comment]
@acumino The API server metrics added in https://github.com/kubernetes/kubernetes/pull/104983 look amazingly useful to me.
Can you open a PR to whitelist these metrics in the Shoot monitoring stack?
https://github.com/gardener/gardener/blob/29a0c4433430ed735e9209c26b106c6817ad45d8/pkg/operation/botanist/component/kubeapiserver/monitoring.go#L184

**Comment by acumino** [comment]
@timebertt I was checking on this 
`consider replacing custom coding in seed-admission-controller with builder.WebhookBuilder.{WithDefaulter,WithValidator}`.
Seems It's not right and we need to follow our old way, which is using `server.Register()`.

To use `builder.WebhookBuilder.{WithDefaulter, WithValidator}`  it needs to implement `admission.{CustomValidator, CustomDefaulter}` [interface](https://github.com/kubernetes-sigs/controller-runtime/blob/5636d975d88e2072884fd82c75b5d3bacf274919/pkg/webhook/admission/validator_custom.go#L31).
If you check, the arguments of the function of that interface it requires a runtime object, and validation/mutation is done on that object.
Also before processing the request, it decodes the object [here](https://github.com/kubernetes-sigs/controller-runtime/blob/5636d975d88e2072884fd82c75b5d3bacf274919/pkg/webhook/admission/validator_custom.go#L73).

But in g/g `seed-admission-controller` case [here](https://github.com/gardener/gardener/blob/eea9eb306639cf7bd9524295bab9582fc452f0f1/pkg/seedadmissioncontroller/webhooks/admission/extensioncrds/admission.go#L77) the `Handle` dependes on multiple resources and not one. So we can't use `builder.WebhookBuilder.{WithDefaulter,WithValidator}`.

**Comment by timebertt** [comment]
Thanks for checking the details.
Regarding `extensioncrds`: you're right the current structure cannot simply be refactored into `admission.CustomValidator`. And probably it doesn't provide much value there. So we can skip this.
Regarding `podschedulername`: this is already simple enough and uses as much upstream logic as possible with `admission.HandlerFunc`.

Regarding `extensionresources`: this code here could be refactored to use `admission.CustomValidator`: https://github.com/gardener/gardener/blob/2beb979582e1c5a98314330046a1e81ca34ec0fa/pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go#L51-L60
I think, this would greatly benefit from it in terms of readability and minimize custom coding.
Basically we can drop this code here in favor of reusing the upstream implementation: https://github.com/gardener/gardener/blob/2beb979582e1c5a98314330046a1e81ca34ec0fa/pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go#L182-L281

**Comment by acumino** [comment]
In case of `extensionresources` also I think the current implementation is simple, since in `extensionresource` also we are dealing with multiple `gvr` if we want to follow  `admission.CustomValidator` we have to implement it for every `gvr` and all of them need to be added to manager separately this will make our code bulkier only. The current approach have a little custom coding but I think this is a better implementation. WDYT?

**Comment by timebertt** [comment]
Hmm, the current approach *does have* quite some custom coding, see the second code block I referenced. Basically half the file can be dropped by reusing upstream implementations.
Setting the webhook for every GVK on the other hand is very simple, doesn't increase maintenance burden and is easy to understand. So I see a clear advantage of refactoring this webhook, do you agree?

**Comment by acumino** [comment]
The implementation will be mostly easy only. But I feel, we are not going to have many benefits from it, I can open a PR in the coming days, and then we can look if we want the newer way.

**Comment by acumino** [comment]
One more hurdle which I missed previously is that the `CustomValidator` returns `error` check [here](https://github.com/kubernetes-sigs/controller-runtime/blob/c46b410039fdc9d620a4217bc7c63c5a7fffae46/pkg/webhook/admission/validator_custom.go#L31-L35) whereas our functions returns `field.ErrorList` check [here](https://github.com/gardener/gardener/blob/f3e414ecd8ff2b0a2143e569b57f93a25a18dc9f/pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go#L75).
So to adapt `CustomValidator` we need to rewrite our validation function i.e. whenever one field is invalid return error at that place only instead of listing down all errors and returning that as `field.ErrorList` which is our current approach.

**Comment by acumino** [comment]
> One more hurdle which I missed previously is that the `CustomValidator` returns `error` check [here](https://github.com/kubernetes-sigs/controller-runtime/blob/c46b410039fdc9d620a4217bc7c63c5a7fffae46/pkg/webhook/admission/validator_custom.go#L31-L35) whereas our functions returns `field.ErrorList` check [here](https://github.com/gardener/gardener/blob/f3e414ecd8ff2b0a2143e569b57f93a25a18dc9f/pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go#L75). So to adapt `CustomValidator` we need to rewrite our validation function i.e. whenever one field is invalid return error at that place only instead of listing down all errors and returning that as `field.ErrorList` which is our current approach.

@timebertt Can you take a look here?

**Comment by timebertt** [comment]
You can call `ToAggregate()` on the `ErrorList`: https://github.com/gardener/gardener/blob/0296476da3a3ee0c4e39acda2f563e09306385f6/vendor/k8s.io/apimachinery/pkg/util/validation/field/errors.go#L241
Then you don't need to change the validation code and can simply call it in the `CustomValidator` implementation.

**Comment by acumino** [comment]
/close
All required tasks have been completed.
